status/views.py

```diff
@@ -0,0 +1,121 @@
+  settings are missing or invalid, to prevent information leakage.
+* Different services provide different information, but all should return
+  "up," DOWN, or NO_CONFIG for the "status" key.
+"""
+
+import re
+from datetime import datetime
+import logging
+
+from django.conf import settings
+from django.http import JsonResponse, Http404
+from elasticsearch import Elasticsearch, ConnectionError as ESConnectionError
+from psycopg2 import connect, OperationalError
+from redis import StrictRedis, ConnectionError as RedisConnectionError
+
+log = logging.getLogger(__name__)
+
+UP = "up"
+DOWN = "down"
+NO_CONFIG = "no config found"
+HTTP_OK = 200
+SERVICE_UNAVAILABLE = 503
+
+
+def get_pg_info():
+    """Check PostgreSQL connection."""
+    try:
+        conf = settings.DATABASES['default']
+    except (AttributeError, KeyError):
+        log.error("No PostgreSQL connection info found in settings.")
+        return {"status": NO_CONFIG}
+    try:
+        start = datetime.now()
+        connection = connect(
+            database=conf["NAME"], user=conf["USER"], host=conf["HOST"],
+            port=conf["PORT"], password=conf["PASSWORD"]
+        )
+        micro = (datetime.now() - start).microseconds
+        connection.close()
+    except (OperationalError, KeyError):
+        log.error("Invalid PostgreSQL connection info in settings: %s", conf)
+        return {"status": DOWN}
+    return {"status": UP, "response_microseconds": micro}
+
+
+def get_redis_info():
+    """Check Redis connection."""
+    pat = re.compile(r'redis://([^:]+):(\d+)/(\d+)$')
+    try:
+        url = settings.BROKER_URL
+    except AttributeError:
+        log.error("No Redis connection info found in settings.")
+        return {"status": NO_CONFIG}
+    try:
+        match = pat.match(url)
+    except TypeError:
+        match = None
+    if match is None:
+        log.error("Invalid Redis connection info found in settings: %s", url)
+        return {"status": NO_CONFIG}
+    host, port, database = match.groups()
+    rdb = StrictRedis(host=host, port=port, db=database)
+    start = datetime.now()
+    try:
+        info = rdb.info()
+    except RedisConnectionError:
+        return {"status": DOWN}
+    micro = (datetime.now() - start).microseconds
+    del rdb  # the redis package does not support Redis's QUIT.
+    ret = {
+        "status": UP, "response_microseconds": micro,
+    }
+    fields = ("uptime_in_seconds", "used_memory", "used_memory_peak")
+    ret.update({x: info[x] for x in fields})
+    return ret
+
+
+def get_elasticsearch_info():
+    """Check Elasticsearch connection."""
+    try:
+        url = settings.HAYSTACK_CONNECTIONS["default"]["URL"]
+    except (AttributeError, KeyError):
+        log.error("No Elasticsearch connection info in settings.")
+        return {"status": NO_CONFIG}
+    search = Elasticsearch(url)
+    start = datetime.now()
+    try:
+        info = search.info()
+    except ESConnectionError:
```

Comment:
I'm also getting a two-minute timeout with `psycopg2` during one test. However, there don't seem to be any timeout kwargs available for any of those libraries.

A quick Google search suggests the best solution is the [signals](https://docs.python.org/2/library/signal.html) package.